kornia/contrib/vit.py

```diff
@@ -233,4 +261,69 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
         out = self.patch_embedding(x)
         out = self.encoder(out)
+        out = self.norm(out)
         return out
+
+    @torch.no_grad()
+    def load_jax_checkpoint(self, checkpoint: str) -> VisionTransformer:
+        import numpy as np
```

Comment:
cc @edgarriba what do you think would be the best approach here?

Reply:
Another approach is to have a function that converts numpy checkpoint to PyTorch's state dict, then the model can just `.load_state_dict()`

Reply:
well, If you think it makes sense to make the effort, we can convert these weights and upload them in[ hf/kornia](https://huggingface.co/kornia)

Reply:
Sure, let's do that.

Reply:
```python
import subprocess
import os

import numpy as np
import torch

from kornia.contrib import VisionTransformer


def convert_jax_checkpoint(np_state_dict: dict[str, np.ndarray]):
    def get_weight(key: str) -> torch.Tensor:
        return torch.from_numpy(np_state_dict[key])

    state_dict = dict()
    state_dict["patch_embedding.cls_token"] = get_weight("cls")
    state_dict["patch_embedding.backbone.weight"] = get_weight("embedding/kernel").permute(3, 2, 0, 1)  # conv weight
    state_dict["patch_embedding.backbone.bias"] = get_weight("embedding/bias")
    state_dict["patch_embedding.positions"] = get_weight("Transformer/posembed_input/pos_embedding").squeeze(0)

    # for i, block in enumerate(self.encoder.blocks):
    for i in range(100):
        prefix1 = f"encoder.blocks.{i}"
        prefix2 = f"Transformer/encoderblock_{i}"

        if f"{prefix2}/LayerNorm_0/scale" not in np_state_dict:
            break

        state_dict[f"{prefix1}.0.fn.0.weight"] = get_weight(f"{prefix2}/LayerNorm_0/scale")
        state_dict[f"{prefix1}.0.fn.0.bias"] = get_weight(f"{prefix2}/LayerNorm_0/bias")

        mha_prefix = f"{prefix2}/MultiHeadDotProductAttention_1"
        qkv_weight = [get_weight(f"{mha_prefix}/{x}/kernel") for x in ["query", "key", "value"]]
        qkv_bias = [get_weight(f"{mha_prefix}/{x}/bias") for x in ["query", "key", "value"]]
        state_dict[f"{prefix1}.0.fn.1.qkv.weight"] = torch.cat(qkv_weight, 1).flatten(1).T
        state_dict[f"{prefix1}.0.fn.1.qkv.bias"] = torch.cat(qkv_bias, 0).flatten()
        state_dict[f"{prefix1}.0.fn.1.projection.weight"] = get_weight(f"{mha_prefix}/out/kernel").flatten(0, 1).T
        state_dict[f"{prefix1}.0.fn.1.projection.bias"] = get_weight(f"{mha_prefix}/out/bias")

        state_dict[f"{prefix1}.1.fn.0.weight"] = get_weight(f"{prefix2}/LayerNorm_2/scale")
        state_dict[f"{prefix1}.1.fn.0.bias"] = get_weight(f"{prefix2}/LayerNorm_2/bias")
        state_dict[f"{prefix1}.1.fn.1.0.weight"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_0/kernel").T
        state_dict[f"{prefix1}.1.fn.1.0.bias"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_0/bias")
        state_dict[f"{prefix1}.1.fn.1.3.weight"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_1/kernel").T
        state_dict[f"{prefix1}.1.fn.1.3.bias"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_1/bias")

    state_dict["norm.weight"] = get_weight("Transformer/encoder_norm/scale")
    state_dict["norm.bias"] = get_weight("Transformer/encoder_norm/bias")
    return state_dict


# recommended checkpoint from https://github.com/google-research/vision_transformer
base_url = "https://storage.googleapis.com/vit_models/augreg"
checkpoint_dict = {
    "vit_l/16": "L_16-i21k-300ep-lr_0.001-aug_strong1-wd_0.1-do_0.0-sd_0.0.npz",
    "vit_b/16": "B_16-i21k-300ep-lr_0.001-aug_medium1-wd_0.1-do_0.0-sd_0.0.npz",
    "vit_s/16": "S_16-i21k-300ep-lr_0.001-aug_light1-wd_0.03-do_0.0-sd_0.0.npz",
    "vit_ti/16": "Ti_16-i21k-300ep-lr_0.001-aug_none-wd_0.03-do_0.0-sd_0.0.npz",
    "vit_b/32": "B_32-i21k-300ep-lr_0.001-aug_light1-wd_0.1-do_0.0-sd_0.0.npz",
    "vit_s/32": "S_32-i21k-300ep-lr_0.001-aug_none-wd_0.1-do_0.0-sd_0.0.npz",
}
variant = "vit_ti/16"

if not os.path.exists(checkpoint_dict[variant]):
    subprocess.run(["wget", f"{base_url}/{checkpoint_dict[variant]}"], check=True)

np_state_dict = np.load(checkpoint_dict[variant])
state_dict = convert_jax_checkpoint(np_state_dict)

# confirm the model can load the weights
vit_model = VisionTransformer.from_config(variant)
vit_model.load_state_dict(state_dict)
```

This is the code to convert to PyTorch's state dict. Change `variant` accordingly to download and convert other variants. You can run it and save the weights, and upload to HF/kornia. Once you uploaded the weights, I will update the code to load the weights from there.

Reply:
Sound good

Reply:
we can rename them if you have better names in mind

- vit_l/16 - https://huggingface.co/kornia/vit_l16_i21k  -- 1210Mb
- vit_b/16 - https://huggingface.co/kornia/vit_b16_i21k -- 343 Mb
- vit_s/16 - https://huggingface.co/kornia/vit_s16_i21k -- 86.7Mb
- vit_ti/16 - https://huggingface.co/kornia/vit_ti16_i21k -- 22.1Mb
- vit_b/32 - https://huggingface.co/kornia/vit_b32_i21k -- 350Mb
- vit_s/32 - https://huggingface.co/kornia/vit_s32_i21k -- 90Mb

Reply:
@johnnv1 These weights actually come from [AugReg](https://arxiv.org/abs/2106.10270), not the original ViT paper, and the [Google's repo](https://github.com/google-research/vision_transformer#available-vit-models) recommends to use these AugReg weights instead. Perhaps we can add the word "augreg" somewhere in the name? (for reference: https://huggingface.co/timm/vit_tiny_patch16_224.augreg_in21k)

Reply:
Also, should we also include the image size in the name? i.e. 224. Similar to https://huggingface.co/kornia/efficientvit_imagenet_b1_r224 and timm models

Reply:
Update the suffix to `..._augreg_i21k_r384` let me know your thoughts (@gau-nernst @edgarriba )...

The model card I believe you can send updates to it as PR on the hugging faces hub later... i just added a minimal info

Reply:
Awesome! Just a small edit. The input size should be 224, not 384. For these AugReg weights, the ImageNet-21k pre-training is 224x224 (which we used). The fine-tuned version on ImageNet-1k is 384x384.

Reply:
ouch, will rename it again

Reply:
https://huggingface.co/kornia/vit_s16_augreg_i21k_r224